hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -429,108 +407,105 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
-        currentElementAddress,
-        destElementType.copyFromType(
-          mb,
-          region,
-          sourceElementType,
-          sourceType.loadElement(srcAddress, numberOfElements, currentIdx),
-          forceDeep
-        )
-      )
+  def copyFromTypeAndStackValue(mb: MethodBuilder, region: Code[Region], srcPType: PType, stackValue: Code[_], forceDeep: Boolean): Code[_] =
+    this.copyFromType(mb, region, srcPType, stackValue.asInstanceOf[Code[Long]], forceDeep)
 
-    if(!sourceElementType.required) {
-      loop = sourceType.isElementMissing(srcAddress, currentIdx).mux(
-        this.setElementMissing(dstAddress, currentIdx),
-        loop
-      )
-    }
+  def constructAtAddress(mb: MethodBuilder, addr: Code[Long], region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Unit] = {
+    val srcArray = srcPType.asInstanceOf[PArray]
 
+    val srcAddrVar = mb.newLocal[Long]
+    val len = mb.newLocal[Int]
     Code(
-      init,
-      Code.whileLoop(currentIdx < numberOfElements,
-        loop,
-        currentElementAddress := this.nextElementAddress(currentElementAddress),
-        currentIdx := currentIdx + const(1)
-      ),
-      dstAddress
+      srcAddrVar := srcAddress,
+      len := srcArray.loadLength(srcAddrVar),
+      Region.storeAddress(addr, constructOrCopyWithLen(mb, region, srcArray, srcAddrVar, len, forceDeep))
     )
   }
 
-  def copyFromType(region: Region, srcPType: PType, srcAddress: Long, forceDeep: Boolean): Long = {
-    val sourceType = srcPType.asInstanceOf[PArray]
-    val sourceElementType = sourceType.elementType.fundamentalType
-    val destElementType = this.elementType.fundamentalType
+  def constructOrCopyWithLen(mb: MethodBuilder, region: Code[Region], srcArray: PArray, srcAddress: LocalRef[Long], len: LocalRef[Int], forceDeep: Boolean): Code[Long] = {
+    if (srcArray == this) {
+      if (forceDeep) {
+        val newAddr = mb.newLocal[Long]
+        Code(
+          newAddr := allocate(region, len),
+          Region.copyFrom(srcAddress, newAddr, contentsByteSize(len)),
+          deepPointerCopy(mb, region, newAddr),
+          newAddr)
+      } else
+        srcAddress
+    } else {
 
-    if (sourceElementType == destElementType) {
-      if(!forceDeep) {
-        return srcAddress
-      }
+      assert(elementType.required <= srcArray.elementType.required)
+
+      val newAddr = mb.newLocal[Long]
+      val i = mb.newLocal[Int]
 
-      return this.deepCopyFromAddress(region, srcAddress)
+      Code(
+        newAddr := allocate(region, len),
+        stagedInitialize(newAddr, len, setMissing = true),
+        i := 0,
+        Code.whileLoop(i < len,
+          srcArray.isElementDefined(srcAddress, i).orEmpty(
+            Code(
+              setElementPresent(newAddr, i),
+              elementType.constructAtAddress(mb, elementOffset(newAddr, len, i), region, srcArray.elementType, srcArray.loadElement(srcAddress, len, i), forceDeep)
```

Comment:
Some inconsistent use of constructAtAddress srcPType field. In some places, called on fundamentalType, in others, like here, on potentially non-fundamental types.